pallets/aggregate/src/lib.rs

```diff
@@ -0,0 +1,827 @@
+                    id,
+                    owner,
+                    next_aggregation_id,
+                    max_aggregation_size,
+                    publish_queue_size,
+                    ticket,
+                )))
+            }
+        }
+
+        /// Compute and reserve the currency for further publication
+        fn reserve_currency_for_publication(
+            &self,
+            account: &AccountOf<T>,
+        ) -> Result<BalanceOf<T>, DispatchError> {
+            let estimated = estimate_publish_aggregation_fee::<T>(self.max_aggregation_size);
+            let hold = (estimated.saturating_add(
+                <T as Config>::ComputePublisherTip::compute_tip(estimated).unwrap_or_default(),
+            )) / self.next.size.into();
+            T::Hold::hold(&HoldReason::Aggregation.into(), account, hold).map(|_| hold)
+        }
+
+        fn take_aggregation(&mut self, id: u64) -> Option<Aggregation<T>> {
+            if self.next.id == id {
+                self.pop_next_aggregation()
+            } else {
+                self.should_publish.remove(&id)
+            }
+        }
+
+        /// Return the size in bytes for this domain that should be reserved in the storage.
+        ///
+        /// - `max_aggregation_size`: The maximum size of the aggregations for this domain.
+        /// - `publish_queue_size`: the publish queue size for this domain.
+        pub fn compute_encoded_size(
+            max_aggregation_size: AggregationSize,
+            publish_queue_size: u32,
+        ) -> usize {
+            DomainType::<T>::compute_encoded_size(max_aggregation_size, publish_queue_size)
+        }
+
+        fn pop_next_aggregation(&mut self) -> Option<Aggregation<T>> {
+            if self.next.statements.is_empty() {
+                None
+            } else {
+                let aggregation = &mut self.next;
+                Some(sp_std::mem::replace(
+                    aggregation,
+                    aggregation.create_next(aggregation.size),
+                ))
+            }
+        }
+
+        fn is_next_aggregation_complete(&self) -> bool {
+            self.next.size as usize <= self.next.statements.len()
+        }
+
+        fn append_statement(
+            &mut self,
+            account: T::AccountId,
+            reserve: BalanceOf<T>,
+            statement: H256,
+        ) -> Option<Aggregation<T>> {
+            self.next.statements.force_push(StatementEntry::new(
+                account.clone(),
+                reserve,
+                statement,
+            ));
+            if self.is_next_aggregation_complete() {
+                self.pop_next_aggregation()
+            } else {
+                None
+            }
+        }
+
+        fn available_aggregation(&mut self, aggregation: Aggregation<T>) {
+            Pallet::<T>::deposit_event(Event::<T>::AggregationComplete {
+                domain_id: self.id,
+                aggregation_id: aggregation.id,
+            });
```

Comment:
I cannot understand your point here. If you you observe this event you can immediately call `aggregate`, the aggregation is complete and you'll receive all founds.

I cannot see any advantage to publish this event in the next block but the logic to do it can be very wired.

Reply:
Ok I got your point but really doesn't mater. If something go wrong the state and events will be reverted, so you cannot see the event.

Anyway I'll move it :smile:

Reply:
Ok, now I see why I didi it :smile: . I move the aggregation in the storage and emit the event need the id... in order to avoid a temp variable to copy id I put it before :smile:

I rollback the change

Reply:
Ok, thanks for clarifying